cmd/kubectl-gadget/deploy.go

```diff
@@ -296,6 +296,8 @@ func runDeploy(cmd *cobra.Command, args []string) error {
 				delete(daemonSet.Spec.Template.Spec.NodeSelector, "kubernetes.io/os")
 			}
 
+			daemonSet.Annotations[utils.GadgetVersionAnnotation] = version
```

Comment:
`version` refers to the kubectl-gadget version in this case. If you deploy something like `kubectl gadget --image=fooo:v0.8.0`, version isn't v0.8.0 but the kubectl-gadget version used to deploy it. Infact, the `INSPEKTOR_GADGET_VERSION` is wrong/misleading. 

https://github.com/kinvolk/inspektor-gadget/blob/585dda7d7924fa2f6830943130cd88044620ea5c/cmd/kubectl-gadget/deploy.go#L317-L318

What about basing the version check on the image tag as I did in https://github.com/kinvolk/inspektor-gadget/pull/936? I know `latest` can be a bit difficult to handle, in that case we can only print this without performing any version comparison. 

Also, it'd be nice to check the versions of all pods and print a different message if there are pods running different versions. (in case the installation is really messed up).

Reply:
> Also, it'd be nice to check the versions of all pods and print a different message if there are pods running different versions. (in case the installation is really messed up).

This should not happen with a single DaemonSet. I think it is simpler to check the DS, also because with the node selector, the gadget pods might only exist on a subset of nodes (e.g. with a mix of Linux & Windows nodes).

Reply:
As far as i understood, when checking the DaemonSet there is still the problem with the version mismatch if someone runs `kubectl gadget --image=fooo:v0.8.0`.
The DaemonSet is going to be tagged with the `clientVersion` even though the image/server version can be a different one.

Reply:
> This should not happen with a single DaemonSet

By "in case the installation is really messed up" I meant cases where the user could be running multiple daemonsets of Inspektor Gadget. I know it's not very probably, but if we're able to print a warning message in that case it would be great. 

> the gadget pods might only exist on a subset of nodes (e.g. with a mix of Linux & Windows nodes).

Is this a problem? I was thinking to get pods by labels (`k8s-app=gadget`) and verify those only.

Reply:
> As far as i understood, when checking the DaemonSet there is still the problem with the version mismatch if someone runs kubectl gadget --image=fooo:v0.8.0.

I think by verifying the daemon set Alban intended to check the image on the daemonset spec:

```
$ kubectl get -n gadget daemonset gadget -o=jsonpath='{.spec.template.spec.containers[0].image}'
```

Reply:
I am wondering if there is a way to make it more useful during the development phase where the image-tag is not the version but the name of the branch:

```bash
$ ./kubectl-gadget version
Client version: v0.8.0-55-g4289e7627b24-dirty
Server version: version
```

We can think/work on it in another PR.

Reply:
It already does that (for me)
```
➭ ./kubectl-gadget version
Client version: 29968d8-dirty
Server version: burak-column-open
```

Or did you mean the branchname should also be in the Client version?

Reply:
kubectl gives more information than just one string:

```
Client Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.2", GitCommit:"9d142434e3af351a628bffee3939e64c681afa4d", GitTreeState:"clean", BuildDate:"2022-01-19T17:35:46Z", GoVersion:"go1.17.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"24", GitVersion:"v1.24.3", GitCommit:"aef86a93758dc3cb2c658dd9657ab4ad4afc21cb", GitTreeState:"clean", BuildDate:"2022-07-13T14:23:26Z", GoVersion:"go1.18.3", Compiler:"gc", Platform:"linux/amd64"}
```

Maybe we could reuse a similar structure with the GitCommit and GitTreeState fields?

Reply:
I created https://github.com/inspektor-gadget/inspektor-gadget/issues/1060 to discuss this.